Source/Core/DolphinQt/Achievements/AchievementProgressWidget.cpp

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#ifdef USE_RETRO_ACHIEVEMENTS
+#include "DolphinQt/Achievements/AchievementProgressWidget.h"
+
+#include <QCheckBox>
+#include <QGroupBox>
+#include <QLabel>
+#include <QLineEdit>
+#include <QProgressBar>
+#include <QPushButton>
+#include <QString>
+#include <QVBoxLayout>
+
+#include <fmt/format.h>
+
+#include <rcheevos/include/rc_api_runtime.h>
+#include <rcheevos/include/rc_api_user.h>
+#include <rcheevos/include/rc_runtime.h>
+
+#include "Core/AchievementManager.h"
+#include "Core/Config/AchievementSettings.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+
+#include "DolphinQt/Config/ControllerInterface/ControllerInterfaceWindow.h"
+#include "DolphinQt/QtUtils/ModalMessageBox.h"
+#include "DolphinQt/QtUtils/NonDefaultQPushButton.h"
+#include "DolphinQt/QtUtils/SignalBlocking.h"
+#include "DolphinQt/Settings.h"
+
+AchievementProgressWidget::AchievementProgressWidget(QWidget* parent) : QWidget(parent)
+{
+  m_common_box = new QGroupBox();
+  m_common_layout = new QVBoxLayout();
+
+  UpdateData();
+
+  m_common_box->setLayout(m_common_layout);
+
+  auto* layout = new QVBoxLayout;
+  layout->setContentsMargins(0, 0, 0, 0);
+  layout->setAlignment(Qt::AlignTop);
+  layout->addWidget(m_common_box);
+  setLayout(layout);
+}
+
+QGroupBox*
+AchievementProgressWidget::CreateAchievementBox(const rc_api_achievement_definition_t* achievement)
+{
+  QLabel* a_title =
+      new QLabel(QString::fromLocal8Bit(achievement->title, strlen(achievement->title)));
+  QLabel* a_description = new QLabel(
+      QString::fromLocal8Bit(achievement->description, strlen(achievement->description)));
+  QLabel* a_points = new QLabel(tr("%1 points").arg(achievement->points));
+  QLabel* a_status = new QLabel(GetStatusString(achievement->id));
+  QProgressBar* a_progress_bar = new QProgressBar();
+  unsigned int value = 0;
+  unsigned int target = 0;
+  AchievementManager::GetInstance()->GetAchievementProgress(achievement->id, &value, &target);
+  if (target > 0)
+  {
+    a_progress_bar->setRange(0, target);
+    a_progress_bar->setValue(value);
+  }
+  else
+  {
+    a_progress_bar->setVisible(false);
+  }
+
+  QVBoxLayout* a_col_right = new QVBoxLayout();
+  a_col_right->addWidget(a_title);
+  a_col_right->addWidget(a_description);
+  a_col_right->addWidget(a_points);
+  a_col_right->addWidget(a_status);
+  a_col_right->addWidget(a_progress_bar);
+  QHBoxLayout* a_total = new QHBoxLayout();
+  // TODO lillyjade 02 Jun 2023 - achievement badge goes here
+  a_total->addLayout(a_col_right);
+  QGroupBox* a_group_box = new QGroupBox();
+  a_group_box->setLayout(a_total);
+  return a_group_box;
+}
+
+void AchievementProgressWidget::UpdateData()
+{
+  QLayoutItem* item;
+  while ((item = m_common_layout->layout()->takeAt(0)) != nullptr)
+  {
+    delete item->widget();
+    delete item;
+  }
+
+  for (u32 ix = 0; ix < AchievementManager::GetInstance()->GetGameData()->num_achievements; ix++)
```

Comment:
Wouldn't creating a local like:

```cpp
const auto* game_data = AchievementManager::GetInstance()->GetGameData();
```

suffice? It'd make the loop and its body a little less verbose by retrieving the data in one spot instead of over and over again.